sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiDynamicDestinationsTableRow.java

```diff
@@ -72,128 +65,86 @@ static void clearSchemaCache() throws ExecutionException, InterruptedException {
-        if (newSchema == null) {
-          throw new RuntimeException("BigQuery table " + tableReference + " not found");
-        }
-        synchronized (this) {
-          tableSchema = newSchema;
-          schemaInformation =
-              TableRowToStorageApiProto.SchemaInformation.fromTableSchema(tableSchema);
-          descriptor = TableRowToStorageApiProto.getDescriptorFromTableSchema(tableSchema);
-          long newHash = BigQueryUtils.hashSchemaDescriptorDeterministic(descriptor);
-          if (descriptorHash != newHash) {
-            LOG.info(
-                "Refreshed table "
-                    + BigQueryHelpers.toTableSpec(tableReference)
-                    + " has a new schema.");
-          }
-          descriptorHash = newHash;
-        }
-      }
+    @Override
+    public com.google.cloud.bigquery.storage.v1.TableSchema getTableSchema() {
+      return protoTableSchema;
+    }
 
-      @Override
-      public TableRow toTableRow(T element) {
-        return formatFunction.apply(element);
-      }
+    @Override
+    public TableRow toTableRow(T element) {
+      return formatFunction.apply(element);
+    }
 
-      @Override
-      public StorageApiWritePayload toMessage(T element) throws Exception {
-        int attempt = 0;
-        do {
-          TableRowToStorageApiProto.SchemaInformation localSchemaInformation;
-          Descriptor localDescriptor;
-          long localDescriptorHash;
-          synchronized (this) {
-            localSchemaInformation = schemaInformation;
-            localDescriptor = descriptor;
-            localDescriptorHash = descriptorHash;
-          }
-          try {
-            Message msg =
-                TableRowToStorageApiProto.messageFromTableRow(
-                    localSchemaInformation,
-                    localDescriptor,
-                    formatFunction.apply(element),
-                    ignoreUnknownValues);
-            return new AutoValue_StorageApiWritePayload(msg.toByteArray(), localDescriptorHash);
-          } catch (SchemaTooNarrowException e) {
-            if (!autoSchemaUpdates || attempt > schemaUpdateRetries) {
-              throw e;
-            }
-            // The input record has fields not found in the schema, and ignoreUnknownValues=false.
-            // It's possible that the user has updated the target table with a wider schema. Try
-            // to read the target's table schema to see if that is the case.
-            refreshSchemaInternal();
-            ++attempt;
-          }
-        } while (true);
-      }
-    };
-  }
+    @Override
+    public StorageApiWritePayload toMessage(T element) throws Exception {
+      return toMessage(formatFunction.apply(element), true);
+    }
+
+    public StorageApiWritePayload toMessage(TableRow tableRow, boolean respectRequired)
+        throws Exception {
+      boolean ignore = ignoreUnknownValues || autoSchemaUpdates;
```

Comment:
autoSchemaUpdates implies ignoreUnknownValues?

I thought user may want autoSchemaUpdates enabled and doesn't want unknown fields?

Reply:
It does not imply that - it is only ignoring unknown values at the prior stage so that it can send them on to the writing stage. However this actually belongs in the followon PR (since in this PR, there is not yet any handling of new schemas) so I'll remove this for now.